packages/core/src/core/runTests.ts

```diff
@@ -14,70 +14,98 @@ export async function runTests(
     rootPath,
     reporters,
     snapshotManager,
+    command,
   } = context;
 
-  const sourceEntries = await getTestEntries({
-    include,
-    exclude,
-    root,
-    fileFilters,
-  });
+  const globTestSourceEntries = async (): Promise<Record<string, string>> => {
+    const entries = await getTestEntries({
+      include,
+      exclude,
+      root,
+      fileFilters,
+    });
+
+    if (!Object.keys(entries).length) {
+      logger.log(color.red('No test files found.'));
+      logger.log('');
+      if (fileFilters.length) {
+        logger.log(color.gray('filter: '), fileFilters.join(color.gray(', ')));
+      }
+      logger.log(color.gray('include:'), include.join(color.gray(', ')));
+      logger.log(color.gray('exclude:'), exclude.join(color.gray(', ')));
+      logger.log('');
+    }
+
+    return entries;
+  };
 
   const setupFiles = getSetupFiles(setups, rootPath);
 
-  if (!Object.keys(sourceEntries).length) {
-    logger.log(color.red('No test files found.'));
-    logger.log('');
-    if (fileFilters.length) {
-      logger.log(color.gray('filter: '), fileFilters.join(color.gray(', ')));
-    }
-    logger.log(color.gray('include:'), include.join(color.gray(', ')));
-    logger.log(color.gray('exclude:'), exclude.join(color.gray(', ')));
-    logger.log('');
-    return;
-  }
+  const rsbuildInstance = await prepareRsbuild(
+    name,
+    globTestSourceEntries,
+    setupFiles,
+  );
+
+  const getRsbuildStats = await createRsbuildServer({
+    name,
+    // TODO: Try not to call globTestSourceEntries again.
+    globTestSourceEntries,
+    setupFiles,
+    rsbuildInstance,
+    rootPath,
+  });
 
-  const rsbuildInstance = await prepareRsbuild(name, sourceEntries, setupFiles);
+  const run = async () => {
+    const buildEnd = Date.now();
+    const testStart = Date.now();
+    const {
+      entries,
+      setupEntries,
+      assetFiles,
+      sourceMaps,
+      getSourcemap,
+      close,
+    } = await getRsbuildStats();
 
-  const { close, entries, assetFiles, sourceMaps, setupEntries, getSourcemap } =
-    await createRsbuildServer({
-      name,
-      sourceEntries,
-      setupFiles,
-      rsbuildInstance,
+    const { results, testResults } = await runInPool({
+      entries,
+      sourceMaps,
+      setupEntries,
+      assetFiles,
+      context,
     });
+    const testEnd = Date.now();
 
-  const buildEnd = Date.now();
+    const duration = {
+      totalTime: testEnd - start,
```

Comment:
shell we need reset start time when watch rerun?

Reply:
Nice catch.